pkgs/stdenv/generic/check-meta.nix

```diff
@@ -86,12 +89,41 @@ let
     allowInsecurePredicate attrs ||
     builtins.getEnv "NIXPKGS_ALLOW_INSECURE" == "1";
 
-  showLicense = license: toString (map (l: l.shortName or "unknown") (lib.lists.toList license));
+  hasSourceProvenance = attrs:
+    attrs ? meta.sourceProvenance;
+
+  isNonSource = sourceTypes: lib.lists.any (t: !t.isSource or true) sourceTypes;
+
+  hasNonSourceProvenance = attrs:
+    hasSourceProvenance attrs &&
+    isNonSource (lib.lists.toList attrs.meta.sourceProvenance);
+
+  # Alow granular checks to allow only some non-source-built packages
+  # Example:
+  # {pkgs, ...}:
+  # {
+  #   allowNonSource = false;
+  #   allowNonSourcePredicate = (x: pkgs.lib.hasPrefix "pulumi" x.name);
+  # }
```

Comment:
I would prefer an example here to allow some source type categorically, ie. `binaryFirmware`. I think that is the more sought after case, and the remediation message when evaluating a disallowed non-source package already contains instructions to do something similar to this.

Reply:
That's quite an ugly thing to do for a mechanism based on the license-handling code because it would involve iterating through a list of source types and duplicating a bunch of logic. It hints at the idea of having a pluggable inner function to determine whether any specific source type should be considered allowed.

But last I checked it's equally ugly to use `allowNonFreePredicate` to categorically allow specific licenses, so if you're interested in fixing this, I'd suggest trying to fix both at the same time.

I'm just interested in having something that's No Worse Than the licenses interface.

Reply:
https://github.com/NixOS/nixpkgs/pull/161098#discussion_r876557725

Reply:
Thanks @a-m-joseph, that is what I am looking for/ is the example that I rather see included here as I imagine it being more helpful to anyone looking there.

I agree that it does not read nicely, but adding helper functions that help with that (and possibly also the license case) can be done in a separate PR after.